metricdog: anonymous bottlerocket metrics

webern opened the pull request:
## Issue Number

Closes #1000 

## Description

Adds a program and systemd service that checks whether certain critical services are running. Reports this to a URL via GET request.

## Testing

### Unit Testing

I organized the code in such a way as to facilitate unit testing everything except the calls to systemctl. 

### Happy Integ

I set up an S3 bucket and confirmed that metricdog requests are working. I ran metricdog manually from the command line under various scenarious and traced the URL that was produced.

#### Healthy Health Ping

```sh
metricdog --log-level trace send-health-ping
```

Result:

```text
https://somedomain.com/metricstest?sender=metricdog&event=health-ping&version=0.4.1&variant=aws-k8s-1.15&arch=x86_64&region=us-west-2&seed=1712&version-lock=latest&ignore-waves=false&failed_services=&is_healthy=true
```

#### Boot Success

```sh
metricdog --log-level trace send-boot-success
```

```text
https://somedomain.com/metricstest?sender=metricdog&event=boot-success&version=0.4.1&variant=aws-k8s-1.15&arch=x86_64&region=us-west-2&seed=1712&version-lock=latest&ignore-waves=false
```

#### One Failed Service

I 'broke' kubelet and ran a health ping, and received:

```text
https://somedomain.com/metricstest?sender=metricdog&event=health-ping&version=0.4.1&variant=aws-k8s-1.15&arch=x86_64&region=us-west-2&seed=1712&version-lock=latest&ignore-waves=false&failed_services=kubelet%3A255&is_healthy=false
```

The failed services list de-escapes to `kubelet:255`

#### Two Failed Services

I 'broke' an additional service:

```text
https://somedomain.com/metricstest?sender=metricdog&event=health-ping&version=0.4.1&variant=aws-k8s-1.15&arch=x86_64&region=us-west-2&seed=1605&version-lock=latest&ignore-waves=false&failed_services=containerd%3A1%2Ckubelet%3A255&is_healthy=false
```

The failed services list de-escapes to `containerd:1,kubelet:255`.

### Bad URL Integ

To make sure that a bad URL doesn't cause problems, I ran an instance with a fake URL. I found that that, send-boot-success and send-health-ping each caused an error to be reported in the system journal, but no other ill affects were observed.

### Systemd

I ran hosts and observed that the systemd unit and timer work as expected.

**Edit 8/28/2020**: I tested an ECS variant with `ab6930c` and found the the URLs are still being constructed properly, including the reporting of failed services.

### Migration

I tested the migration through upgrade and downgrade, confirming that settings were added/removed and the system was healthy at all three waypoints.

## Terms of contribution:

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

tjkirch commented on sources/models/src/aws-k8s-1.15/mod.rs:
```diff
@@ -3,7 +3,7 @@ use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
```
We'll need the change and settings for the aws-ecs-1 model as well.  And I'd argue for aws-dev, too.

webern replied:
Addressed (and tested) in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

tjkirch commented on packages/os/healthdog-toml:
```diff
@@ -0,0 +1,7 @@
+metrics_url = "{{settings.healthdog.metrics-url}}"
+send_metrics = {{settings.healthdog.send-metrics}}
+service_health = [{{#each settings.healthdog.service-health as |service| ~}}"{{service}}",{{/each~}}]
```
What is this doing?  I don't think the syntax is obvious and I can't find documentation on it.  Would it be clearer to use something like the `join_map` we have today, but for a list?

webern replied:
This is, indeed, doing a join to surround each member of the vector of strings with quotes and a trailing comma. Is `join_map` a feature of the Handlebars crate? Or are you saying there is an example of how to handle a vector that exists in our code? (in Schnauzer?)

tjkirch replied:
Yeah, schnauzer has a helpers module with things like `join_map`, `default`, etc.

webern replied:
It doesn't look like there is an existing helper in schnauzer for this, but the handlebars crate comes out-of-the-box with the ability to write this as I have here.

tjkirch replied:
I think we could make a simple helper that was much clearer than that handlebars syntax, though...

webern replied:
Created the helper in [f515d38](https://github.com/bottlerocket-os/bottlerocket/compare/d62f75d5060aa2f3ac38e7a7f0a7d25d29811b62..f515d383c9e181c8a531a02414b320ff9231a81f)

tjkirch commented on packages/os/healthdog.service:
```diff
@@ -0,0 +1,7 @@
+
+[Service]
+Type=oneshot
+RemainAfterExit=false
```
Why false?  We usually have this be true so that the service doesn't look unhealthy if it ran OK the last time.

webern replied:
`true` definitely did not work with the timer. I don't pretend to be a systemd expert, but I think RemainAfterExit makes it seem like the service is 'active' and the timer doesn't start it again.

tjkirch replied:
Ah, ok, I don't have much experience with timers :)

tjkirch commented on packages/os/healthdog.service:
```diff
@@ -0,0 +1,7 @@
+[Service]
+Type=oneshot
+RemainAfterExit=false
+ExecStart=/bin/healthdog send-health-ping
```
```suggestion
ExecStart=/usr/bin/healthdog send-health-ping
```

webern replied:
Does this mean I should change the install directives as well in os.spec?

tjkirch replied:
In spec files you'd be using macros, so "/bin" vs. "/usr/bin" shouldn't appear anyway.

webern replied:
I didn't realize the program could be called using either path, but it seems to work with no changes to os.spec.

Addressed in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

tjkirch commented on packages/os/healthdog.timer:
```diff
@@ -0,0 +1,15 @@
+# Don't run missed executions
+Persistent=false
+# Run 120 seconds after boot for the first time
+OnBootSec=120
```
```suggestion
OnStartupSec=120
```
This is relative to the init system rather than the hardware boot, which seems like a more logical thing to think about for me - if an initrd or kernel startup takes longer, it doesn't mean we want to give less time for services to start before checking them.

webern replied:
Addressed in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

tjkirch commented on packages/os/healthdog.timer:
```diff
@@ -0,0 +1,15 @@
+# Run 120 seconds after boot for the first time
+OnBootSec=120
+# Run every 6 hours thereafter
+OnUnitActiveSec=21600
```
```suggestion
OnUnitActiveSec=21600
# Don't fire at exactly the same second across machines started together.
RandomizedDelaySec=10
# We don't want to extend the startup report too long after the requested time.
AccuracySec=10
```

webern replied:
Addressed in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

tjkirch commented on sources/models/defaults.toml:
```diff
@@ -52,6 +52,24 @@ template-path = "/usr/share/templates/updog-toml"
+
+[settings.healthdog]
+# the URL to which anonymous health metrics will be sent
+metrics-url = "https://metrics.bottlerocket.aws/metrics"
```
```suggestion
metrics-url = "https://metrics.bottlerocket.aws"
```
We don't need this, do we?

webern replied:
Addressed in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

tjkirch commented on sources/models/defaults.toml:
```diff
@@ -52,6 +52,24 @@ template-path = "/usr/share/templates/updog-toml"
+# whether or not health metrics will be sent. set to false to opt-out
+send-metrics = true
+# the list of services that are checked to determine if a host is healthy
+service-health = ["apiserver"]
```
"service-health" isn't very clear to me. When I first saw it in the model, I assumed you were using it as a storage location for service health check results, and I was confused. How about "check-services" or similar?

webern replied:
Sure, or just "services"?

webern replied:
Addressed in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

tjkirch commented on sources/models/src/aws-k8s-1.15/override-defaults.toml:
```diff
@@ -36,3 +36,9 @@ affected-services = ["kubernetes", "containerd"]
+# Healthdog
+
+[settings.healthdog]
+metrics-url = "https://metrics.bottlerocket.aws/metrics/"
```
```suggestion
```

I don't think we have to override this just to override the other field?

tjkirch reviewed (commented):
Some comments from the surrounding bits; still need to review healthdog itself.

samuelkarp commented on sources/healthdog/README.md:
```diff
@@ -0,0 +1,55 @@
+* `version`: the Bottlerocket version.
+* `variant`: the Bottlerocket variant.
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
+* `region`: the region the machine is running in.
```
It might be worth calling this `aws-region` instead to make it more clear that this is an AWS-specific data field.

webern replied:
Addressed in [f18d841](https://github.com/bottlerocket-os/bottlerocket/compare/ec0abbf943fe3806294f53fd4e118be3d1f529fc..f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae)

samuelkarp replied:
It looks like it still says `region` on this line?

webern replied:
That is for the query params, the config says aws_region. Edit: Tom is requesting that I change the query param to aws-region as well.

tjkirch replied:
Yeah; I'm assuming we'll have similarly named parameters for other platforms, so we'll want to be able to distinguish from a single field inside the query params as well.  (variant will tell us platform too, probably, but this is the one-stop shop.)

webern replied:
Fixed in [1083d9b](https://github.com/bottlerocket-os/bottlerocket/compare/3fae608ba56a631a4d63129c1c854916d2dd61b6..1083d9b81122e5e5bd42940472cc6201d8bd5149)

tjkirch commented on sources/healthdog/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+[[bin]]
+name = "healthdog"
+path = "src/main.rs"
+
```
```suggestion
```
I don't think you need this if you just have one binary at `src/main.rs`, named the same as the crate?

webern replied:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

tjkirch commented on sources/healthdog/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+path = "src/main.rs"
+
+[dependencies]
+bottlerocket-release = { version = "0.1.0", path = "../bottlerocket-release"}
```
```suggestion
bottlerocket-release = { path = "../bottlerocket-release"}
```
We need to keep it in sync, and don't have plans to publish it separately, so the version shouldn't matter.

webern replied:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

tjkirch commented on sources/healthdog/src/main_test.rs:
```diff
@@ -0,0 +1,208 @@
+use std::path::PathBuf;
+use tempfile::TempDir;
+
+const OS_RELEASE: &str = r#"PRETTY_NAME=Bottlerocket,
```
```suggestion
const OS_RELEASE: &str = r#"PRETTY_NAME=Bottlerocket
```
Guessing that was a typo?

webern replied:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

tjkirch commented on sources/healthdog/src/main_test.rs:
```diff
@@ -0,0 +1,208 @@
+// dynamically create a config file where we can set server port, list of services, and send_metrics
+fn create_config_file_contents(port: u16, services: &[&str], send_metrics: bool) -> String {
+    let mut svcs = String::new();
+    for (i, &service) in services.iter().enumerate() {
+        if i == 0 {
+            svcs.push('"')
+        }
+        svcs.push_str(service);
+        if i == services.len() - 1 {
+            svcs.push('"');
+        } else {
+            svcs.push_str("\",\"");
+        }
+    }
```
Could this be simplified to something like `format!("\"{}\"", services.join(","))`?

webern replied:
Not quite, but this will work (will push soon):

```rust
    let svcs = services
        .iter()
        .map(|&s| format!("\"{}\"", s))
        .collect::<Vec<String>>()
        .join(", ");
```

webern replied:
Implemented the more functional version in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

webern commented:
> webern force-pushed the webern:healthdog branch from ec0abbf to f18d841

* renamed `settings.healthdog` to `settings.metrics`.
* renamed `region` to `aws-region`.
* added (and tested) settings for ecs and dev variants.
* made suggested changes to systemd configs (and tested them)

tjkirch commented on sources/models/src/aws-dev/override-defaults.toml:
```diff
@@ -1,3 +1,9 @@
+
+[settings.metrics]
+send-metrics = false
+services = []
```
Why not send metrics about at least the apiserver?  I think we could leave this to the defaults?

bcressey replied:
Maybe also chrony, containerd, host-containerd, since they're all related to default settings.

chrony could easily go unhealthy as a result of a chrony or libc update that changes the system calls it makes, causing the seccomp filter to kill it.

webern replied:
Added in [11d0d7b](https://github.com/bottlerocket-os/bottlerocket/compare/cbc5d1d0b587442093072c22e97e1a9398d9f153..11d0d7bdd62c0b6555729a5854732b295227af9e)

tjkirch commented on sources/healthdog/src/healthdog_test.rs:
```diff
@@ -0,0 +1,162 @@
+        Some(Config {
+            metrics_url: format!("http://localhost:{}/metrics", port),
+            send_metrics: true,
+            // note that these are out-of-order sort order to ensure that failed services are sorted
+            // in the url
```
Does it matter if they're sorted?

webern replied:
I think so. If one ever needed to pour over raw records at 3 AM, it would be a lot easier to spot a pattern. The cost of the sort is infinitesimal so I would vote for leaving the sort which produces deterministic urls.

tjkirch commented on sources/healthdog/src/args.rs:
```diff
@@ -0,0 +1,197 @@
+            "--help" | "-h" => return Err(Error::Usage { message: None }),
+            // Assume any argument not prefixed with '-' is a subcommand
+            s if !s.starts_with('-') => {
+                ensure!(
```
As much as I like good, simple arg parsing code, when you get into subcommands I start to question the benefit.  You may want to take a look at argh, which Ben has used recently; that turned out well.  Most also allow specifying TryFrom/FromStr or a parsing function, and handling their resulting errors, rather than doing it manually like the bit above for LevelFilter.

webern replied:
Introduced structopt in [a58c5e2](https://github.com/bottlerocket-os/bottlerocket/compare/9dc047c8079357f0e3c547672601e66d78a376b5..a58c5e2b6e5782bd16acaef0134ed481ce32cde4).

I tried argh, but it was a bit ugly for what I wanted it to do, structopt came out much cleaner.

tjkirch commented on sources/healthdog/src/args.rs:
```diff
@@ -0,0 +1,197 @@
+    fn parse<S: AsRef<str>>(s: S) -> Result<Self> {
+        match s.as_ref() {
+            BOOT_SUCCESS => Ok(Command::BootSuccess),
+            HEALTH_PING => Ok(Command::HealthPing),
```
If you don't want to try a parser (I suggested argh elsewhere) then I'd recommend renaming the enum variants to match the user-visible names so you could use something like serde_plain here, rather than constants and matching.

webern replied:
Fixed in [a58c5e2](https://github.com/bottlerocket-os/bottlerocket/compare/9dc047c8079357f0e3c547672601e66d78a376b5..a58c5e2b6e5782bd16acaef0134ed481ce32cde4)

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+* `version`: the Bottlerocket version.
+* `variant`: the Bottlerocket variant.
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
+* `region`: the region the machine is running in.
```
```suggestion
* `aws-region`: the AWS region the machine is running in.
```

webern replied:
Fixed in [1083d9b](https://github.com/bottlerocket-os/bottlerocket/compare/3fae608ba56a631a4d63129c1c854916d2dd61b6..1083d9b81122e5e5bd42940472cc6201d8bd5149)

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
+* `region`: the region the machine is running in.
+* `seed`: the seed value used to roll-out updates.
+* `version-lock`: the optional setting that locks Bottlerocket to a certain version.
```
```suggestion
* `version-lock`: the optional setting that controls Bottlerocket update selection.
```
Not necessarily a certain version, it can be "latest" (and maybe other things in the future)

webern replied:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+use std::{env, process};
+
+fn main() -> ! {
+    process::exit(match main_inner(env::args(), Box::new(SystemdCheck {})) {
```
```suggestion
    process::exit(match main_inner(env::args(), Box::new(SystemdCheck)) {
```
Don't need empty `{}` for structs with no fields.

webern replied:
Seemingly not possible?

```
81 |     process::exit(match main_inner(args, Box::new(SystemdCheck)) {
   |                                                   ^^^^^^^^^^^^ did you mean `SystemdCheck { /* fields */ }`?
```

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+use std::{env, process};
+
+fn main() -> ! {
+    process::exit(match main_inner(env::args(), Box::new(SystemdCheck {})) {
```
I'd prefer to stick to the standard `Result`-based structure used in most of our binaries, rather than this manual exit and return-code handling, unless there's some reason I'm missing...

webern replied:
Seemingly not true?

```
81 |     process::exit(match main_inner(args, Box::new(SystemdCheck)) {
   |                                                   ^^^^^^^^^^^^ did you mean `SystemdCheck { /* fields */ }`?
```

webern replied:
Perhaps it looks better now that I'm using structopt? [a58c5e2](https://github.com/bottlerocket-os/bottlerocket/compare/9dc047c8079357f0e3c547672601e66d78a376b5..a58c5e2b6e5782bd16acaef0134ed481ce32cde4)

The reason for matching on the Result is to use snafu's implementation of the Display trait to display the erro message. This looks like many of the main functions I've seen in our codebase (modulo a possible stylistic tweak).

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+}
+
+/// To facilitate testing of `main_inner`, ensure that the logger is only initialized once.
+static INIT_LOGGER_ONCE: Once = Once::new();
```
Can we parse args and initialize the logger in `main` instead of needing this extra `Once`?

webern replied:
Fixed when structopt was introduced: [a58c5e2](https://github.com/bottlerocket-os/bottlerocket/compare/9dc047c8079357f0e3c547672601e66d78a376b5..a58c5e2b6e5782bd16acaef0134ed481ce32cde4)

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+    match arguments.command {
+        Command::BootSuccess => {
+            if let Err(err) = healthdog.send_boot_success() {
+                // we don't want to fail the boot if there is a failure to send this message, so
+                // we log the error and return Ok(())
```
We *are* failing if `send_health_ping` fails, though - just to be sure, the timer will still run again after that, right?

webern replied:
I've confirmed that the timer works like we want it to. I set the timer to run every 20 seconds for testing, and I see it running every 20 seconds even when failing.

```
Aug 28 00:39:08 ip-10-192-10-222.us-west-2.compute.internal systemd[1]: healthdog.service: Failed with result 'exit-code'.
Aug 28 00:39:28 ip-10-192-10-222.us-west-2.compute.internal systemd[1]: healthdog.service: Failed with result 'exit-code'.
Aug 28 00:39:58 ip-10-192-10-222.us-west-2.compute.internal systemd[1]: healthdog.service: Failed with result 'exit-code'.
Aug 28 00:40:18 ip-10-192-10-222.us-west-2.compute.internal systemd[1]: healthdog.service: Failed with
...etc
```

I've also confirmed that when `send-boot-success` fails, we do not cause `mark-successful-boot.service` to report a failure.

```
Aug 28 00:37:34 ip-10-192-10-222.us-west-2.compute.internal healthdog[3911]: [2020-08-28T00:37:34Z ERROR healthdog] Error while reporting boot success: Error sending HTTP request to https://metrics.bottlerocket.aws/?sender=healthdog&event=boot_success&version=1.0.0&variant=aws-ecs-1&arch=x86_64&aws_region=us-west-2&seed=1610&version_lock=latest&ignore_waves=false: error sending request for url (https://metrics.bottlerocket.aws/?sender=healthdog&event=boot_success&version=1.0.0&variant=aws-ecs-1&arch=x86_64&aws_region=us-west-2&seed=1610&version_lock=latest&ignore_waves=false): error trying to connect: dns error: failed to lookup address information: Name or service not known
```

`systemctl status mark-successful-boot` reports successful and healthdog did exit 0:

```
 Process: 3911 ExecStartPost=/usr/bin/healthdog --log-level trace send-boot-success (code=exited, status=0/SUCCESS)
```

tjkirch commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+    if !config.send_metrics {
+        return Ok(());
+    }
+    let os_release = if let Some(os_release_path) = &arguments.os_release {
+        BottlerocketRelease::from_file(os_release_path)
+    } else {
+        BottlerocketRelease::new()
+    }
```
Is there a benefit to doing the defaulting of healthdog parts here *and* in `from_parts`?

webern replied:
fixed in the `3909ee7` push, but the diff got messed up

I was doing this twice because I was imagining using healthdog as a library. But since it's not a library yet, there was no point. Also if I'm the one to use healthdog as a library, I will still probably want to inject my dependencies anyway.

tjkirch commented on sources/healthdog/src/healthdog.rs:
```diff
@@ -0,0 +1,160 @@
+        S2: AsRef<str>,
+    {
+        let mut url = Url::from_str(&self.config.metrics_url).context(error::UrlParse {
+            url: self.config.metrics_url.clone(),
```
Can we parse the URL earlier, when constructing the config?

webern replied:
Not easily when constructing config since Url does not implement the deserialize trait. Also if you opt-out then config shouldn't care about the Url. So what I did in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306) is construct the Url when Healthdog is constructed, which happens after opt-out has been checked.

tjkirch commented on sources/healthdog/src/healthdog.rs:
```diff
@@ -0,0 +1,160 @@
+            q.append_pair("version", &version);
+            q.append_pair("variant", &self.os_release.variant_id);
+            q.append_pair("arch", &self.os_release.arch);
+            q.append_pair("region", &self.config.aws_region);
```
```suggestion
            q.append_pair("aws_region", &self.config.aws_region);
```

webern replied:
this is fixed in [1083d9b](https://github.com/bottlerocket-os/bottlerocket/compare/3fae608ba56a631a4d63129c1c854916d2dd61b6..1083d9b81122e5e5bd42940472cc6201d8bd5149)

tjkirch commented on sources/healthdog/src/healthdog.rs:
```diff
@@ -0,0 +1,160 @@
+            q.append_pair("sender", sender.as_ref());
+            q.append_pair("event", event.as_ref());
+            let version = format!("{}", &self.os_release.version_id);
+            q.append_pair("version", &version);
```
```suggestion
            q.append_pair("version", self.os_release.version_id.to_string());
```
Unless this specifically needs to be a reference, but `to_string` is still simpler/clearer than `format`, I think.

Same below for ignore-waves.

webern replied:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

tjkirch commented on sources/healthdog/src/healthdog.rs:
```diff
@@ -0,0 +1,160 @@
+                }
+            }
+        }
+        Self::send_get_request(url, timeout_seconds.unwrap_or(DEFAULT_TIMEOUT_SECONDS))?;
```
Might be simpler to handle the `.unwrap_or(DEFAULT_TIMEOUT_SECONDS)` inside `send_get_request` than in all callers?

webern replied:
fixed in the `3909ee7` push, but the diff got messed up

tjkirch commented on sources/healthdog/src/healthdog.rs:
```diff
@@ -0,0 +1,160 @@
+    /// services, and `1` and `2` are exit codes of the failed services.
+    pub(crate) fn send_health_ping(&self) -> Result<()> {
+        let mut is_healthy = true;
+        let mut failed_services: Vec<String> = Vec::new();
```
```suggestion
        let mut failed_services = Vec::new();
```

webern replied:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

tjkirch commented on sources/healthdog/src/service_check.rs:
```diff
@@ -0,0 +1,127 @@
+}
+
+/// Regex pattern for finding the exit code of a systemd service that has exited. There is a single
+/// capture group, named `exit_code`.
```
```suggestion
/// capture group, named `exit_code`.  Example text:
///     Main PID: 1107 (code=exited, status=17)
```

webern replied:
This went away with [11d0d7b](https://github.com/bottlerocket-os/bottlerocket/compare/cbc5d1d0b587442093072c22e97e1a9398d9f153..11d0d7bdd62c0b6555729a5854732b295227af9e)

tjkirch commented on sources/healthdog/src/service_check.rs:
```diff
@@ -0,0 +1,127 @@
+/// Regex pattern for finding the exit code of a systemd service that has exited. There is a single
+/// capture group, named `exit_code`.
+const SYSTEMD_EXIT_REGEX_PATTERN: &str =
+    r#"Main PID: \d+ \(code=[a-zA-Z0-9-_]+, status=(?P<exit_code>\d{1,3})/[a-zA-Z0-9-_]+\)"#;
```
I think the "/" part after the code is optional; at least, I'm not seeing it in an example on my machine.

In any case, I'd really like to avoid parsing this if we can.  I believe they consider this output for humans and could change it at any time.  `systemctl show` has a much simpler output format than `systemctl status`; it seems to be intended for machines.  There's a line like `ExecMainStatus=17` that would, obviously, be easier to parse; I wouldn't even use a regex.  Would that work for us?

webern replied:
Huuuuuuge improvement in [11d0d7b](https://github.com/bottlerocket-os/bottlerocket/compare/cbc5d1d0b587442093072c22e97e1a9398d9f153..11d0d7bdd62c0b6555729a5854732b295227af9e). Thank you!

tjkirch commented on sources/healthdog/src/service_check.rs:
```diff
@@ -0,0 +1,127 @@
+}
+
+fn is_ok(service: &str) -> Result<bool> {
+    Ok(!is_failed(service)? && is_active(service)?)
```
Is is-active always safe?  What if we wanted to check that a oneshot service was successful?

webern replied:
I think I agree with Jamie on this:

> It'd be nice to eventually add support for non-lingering oneshot services, but I think this is a good place to start.

The benefit of checking is-active is that we would know if a service never started (as opposed to started and failed). For now, supporting a oneshot service with RemainAfterExit=false seems like YAGNI. But if we do end up needing it in the future, I can imagine ways of adding supporting for it in a backwards compatible way.

tjkirch reviewed (commented):
Not quite done with service_check, but have to stop for the moment...

tjkirch commented on sources/healthdog/src/service_check.rs:
```diff
@@ -0,0 +1,127 @@
+        }
+        Ok(ServiceHealth {
+            is_healthy: false,
+            exit_code: parse_service_exit_code(service_name)?,
```
If we can't parse it, should we return None rather than failing the whole check?  It's possible a systemd update breaks this module, and we'd want to know...

webern replied:
Yeah good point, fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)

etungsten commented on sources/healthdog/src/main.rs:
```diff
@@ -0,0 +1,138 @@
+### Additionally, when `healthdog` sends a 'health ping', it adds:
+
+* `is-healthy`: true or false based on whether critical services are running.
+* `failed_services`: a list of critical services that have failed, if any.
```
To keep underscores/hyphens consistent.
```suggestion
* `failed-services`: a list of critical services that have failed, if any.
```

Or the other way around. Maybe these should all be snake_cased to reflect how the query parameters are structured.

webern replied:
More confusing than I expected, trying to sort this out based on the following:

* config: e.g. healthdog-toml should use snake_case to be consistent with updog-toml.
* query parameters: snake_case because we can choose what we want and more things are snake than kebob in our world.
* settings: e.g. defaults.toml, kebob-case to be consistent with what's there now.

webern replied:
Normalized to snake_case in [1083d9b](https://github.com/bottlerocket-os/bottlerocket/compare/3fae608ba56a631a4d63129c1c854916d2dd61b6..1083d9b81122e5e5bd42940472cc6201d8bd5149)

bcressey commented on sources/models/defaults.toml:
```diff
@@ -52,6 +52,24 @@ template-path = "/usr/share/templates/updog-toml"
 
+# Metrics
+
+[settings.metrics]
+# the URL to which anonymous health metrics will be sent
+metrics-url = "https://metrics.bottlerocket.aws"
+# whether or not health metrics will be sent. set to false to opt-out
+send-metrics = true
+# the list of services that are checked to determine if a host is healthy
+services = ["apiserver"]
```
I'm late to the party but `settings.metrics` feels a bit general; it could be taken to mean performance metrics as well, if we gathered those.

A couple options to consider:
* rename `send-metrics` to `send-health-metrics` with the thought that other types of metrics would get their own setting
* keep `send-metrics` as-is and treat it as an all-or-nothing way to control sending metrics

The second option seems fine to me; anyone who takes the time to disable this won't want to discover that there's another setting that they need to turn off in a later version.

Assuming we take that path, then something like `service-checks` might be better than `services`. The downside to `services` is that it might seem like we're collecting more metrics than we really are for containerd et al.

webern replied:
I like the second option and will rename services to service-checks unless i hear otherwise.

webern replied:
I changed the name of the setting from `services` to `service-checks` in [cbc5d1d](https://github.com/bottlerocket-os/bottlerocket/compare/a58c5e2b6e5782bd16acaef0134ed481ce32cde4..cbc5d1d0b587442093072c22e97e1a9398d9f153).

bcressey commented on sources/models/src/aws-ecs-1/override-defaults.toml:
```diff
@@ -13,3 +13,8 @@ affected-services = ["ecs"]
+# Metrics
+
+[settings.metrics]
+services = ["containerd", "docker", "ecs"]
```
nit: a few of these have the "no newline at EOF" warning marker

webern replied:
Pretty sure these got fixed up in rebase [4401e25](https://github.com/bottlerocket-os/bottlerocket/compare/f18d841ee1c2b35d5f09136ea3f7136e2a4d65ae..4401e252719fbec5532cbec58c0b43b15f4e3fa6)

tjkirch commented:
> Maybe also chrony, containerd, host-containerd, since they're all related to default settings.

host-containerd sure, but containerd might not be used in all variants..?

webern committed:
metricdog: anonymous bottlerocket metrics

Metricdog is a new command-line program that sends anonymous information
about a Bottlerocket instance to a metrics endpoint. A report of boot
success is sent by the mark-boot-success service and at 2 minutes after
boot, and every 6 hours thereafter, a report of service health is sent.

webern commented:
> webern force-pushed the webern:healthdog branch from f18d841 to 4401e25

First attempt at a slightly precarious rebase.

webern commented:
>  webern force-pushed the webern:healthdog branch from 4401e25 to 809f98f

Fix horrible things from the first rebase attempt.

webern commented:
> webern force-pushed the webern:healthdog branch from 809f98f to 3fae608

Undo a tiny re-order that was a rebase artifact.

webern commented:
> webern force-pushed the webern:healthdog branch from 3fae608 to 1083d9b

Use aws_region instead of region and normalize on snake_case for params.

webern commented:
> webern force-pushed the webern:healthdog branch from 1083d9b to d62f75d

Rebase the swift-moving develop again.

webern commented:
> webern force-pushed the webern:healthdog branch from d62f75d to f515d38

Create a `join-strings` helper in schnauzer to simplify the `healthdog-toml` template.

webern commented:
> webern force-pushed the webern:healthdog branch from f515d38 to 2a5aef0

Fixes most of the lowest-hanging fruit in the PR comments.

webern commented:
> webern force-pushed the webern:healthdog branch from 2a5aef0 to 9dc047c

Weeee... rebase the migration versions again from develop! :)

webern commented:
> webern force-pushed the webern:healthdog branch from 9dc047c to a58c5e2

Use `structopt`.

webern commented:
> webern force-pushed the webern:healthdog branch from a58c5e2 to cbc5d1d 

Rename the `services` setting to `service-checks`.

webern commented:
> > Maybe also chrony, containerd, host-containerd, since they're all related to default settings.
> 
> host-containerd sure, but containerd might not be used in all variants..?

It's in all current variants. I think it would be good to add these to the top-level defaults.toml. A variant that doesn't use containerd can still override.

P.S. We're actually overriding for all current variants now (on my WIP branch), so `defaults.toml` more like an example anyway.

webern commented:
> webern force-pushed the webern:healthdog branch from cbc5d1d to 11d0d7b

* Enormous improvement to the systemctl exit code parsing (thank you, Tom)
* Add containerd and chronyd to the default service checks
* fix some issues with my renaming of the service-checks setting

webern commented on sources/healthdog/src/error.rs:
```diff
@@ -0,0 +1,51 @@
+    HttpResponse { url: Url, source: reqwest::Error },
+
+    #[snafu(display("Usage error."))]
+    Usage { message: Option<String> },
```
Remove any unused errors before requesting re-review.

jamieand reviewed (approved):
Looking specifically at the interactions with systemd:

It'd be nice to eventually add support for non-lingering oneshot services, but I think this is a good place to start.

webern commented:
> webern force-pushed the webern:healthdog branch from 328b475 to 3909ee7

Fix remaining issues, esp. providing construction defaults in two places.

webern commented:
> webern force-pushed the webern:healthdog branch from 3909ee7 to ab6930c

Fix a bug with the exit code parsing.

webern commented:
> webern force-pushed the webern:healthdog branch from ab6930c to eaefea7

Delete an unused error.

webern commented:
> webern force-pushed the webern:healthdog branch from eaefea7 to b5f2835

git rebase develop

webern commented:
> webern force-pushed the webern:healthdog branch from b5f2835 to aa39d29 

rebase

webern commented:
> webern force-pushed the webern:healthdog branch from aa39d29 to d2e6d7b

rebase and bump migration version

webern commented:
> webern force-pushed the webern:healthdog branch from d2e6d7b to 0fc617f

rebase

webern commented:
> webern force-pushed the webern:healthdog branch from 0fc617f to 153acff

Rename to `metricdog`.

Not sure what this does to our ability to follow the PR comments and files that they apply to, but the rename is done.
I don't think I can rename the branch without harming the PR. LMK if I'm wrong.

tjkirch commented on packages/os/mark-successful-boot.service:
```diff
@@ -8,6 +8,7 @@ After=multi-user.target
 Type=oneshot
 RemainAfterExit=true
 ExecStart=/bin/signpost mark-successful-boot
+ExecStartPost=/usr/bin/metricdog send-boot-success
```
I think we may want the `-` prefix so that service startup continues if we can't send a metric ping. (https://man7.org/linux/man-pages/man5/systemd.service.5.html)

webern replied:
Added in [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

tjkirch commented on sources/Cargo.toml:
```diff
@@ -28,11 +28,14 @@ members = [
     "api/migration/migrations/v1.0.0/ecr-helper-admin",
     "api/migration/migrations/v1.0.0/ecr-helper-control",
     "api/migration/migrations/v1.0.2/add-enable-spot-instance-draining",
+    "api/migration/migrations/v1.0.2/metricdog-init",
```
May want to make this v1.0.3 preemptively.  (and the directory, etc.)

tjkirch commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -463,6 +475,65 @@ pub fn ecr_prefix(
+/// `{{ join-strings settings.somewhere.foo-list }}`
+///
+/// This will render `"a", "b", "c"`.
+pub fn join_strings(
```
Maybe `join_array` for parity with the existing `join_map`?

webern replied:
Implemented in [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737)

tjkirch commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -463,6 +475,65 @@ pub fn ecr_prefix(
+///
+/// Consider an array of values: `[ "a", "b", "c" ]` stored in a setting such as
+/// `settings.somewhere.foo-list`. In our template we can write:
+/// `{{ join-strings settings.somewhere.foo-list }}`
```
Should we have a first parameter for the joining string, like `join_map`?  It might not always be a comma, and I think people are used to specifying that character with join functions.

webern replied:
Implemented in [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737)

tjkirch commented on sources/models/defaults.toml:
```diff
@@ -52,6 +52,24 @@ template-path = "/usr/share/templates/updog-toml"
+metrics-url = "https://metrics.bottlerocket.aws"
+# whether or not health metrics will be sent. set to false to opt-out
+send-metrics = true
+# the list of services that are checked to determine if a host is healthy
```
```suggestion
# the list of services that are checked to determine if a host is healthy, overridden in each variant to list services critical to that variant
```
I do like having a default here just in case, but I also think it'd be nice to clarify that we expect it to be overridden.

webern replied:
Added in [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

tjkirch commented on sources/metricdog/src/main.rs:
```diff
@@ -0,0 +1,121 @@
+
+fn main() -> ! {
+    let args = Arguments::from_args();
+    EnvLogger::new()
```
Is there a reason to veer away from simplelog like we use in all our other crates?  We're talking about changing our logger, but that's easier if we're consistent.

webern replied:
At first I copied code from updog which panicked when systemd ran the program. Looking deeper now it appears maybe I could have instantiated a SimpleLogger(?) instead of TermLogger, but didn't figure that out at the time. Seems like simplelog has caused issues for us elsewhere as well.

`env_logger` is more popular, is highly user-configurable (at runtime), seems to have fewer foot-guns and seems like the closest thing to a de-facto standard (though there are way, way too many logger crates).

What I like about env_logger is that, as the user at runtime you can control the logging verbosity of even dependent crates. We should probably choose a logger and standardize on it, but I wouldn't choose simplelog for the standard. I'd be happy to replace simplelog with env_logger throughout the codebase (in a separate PR) if we want to get that done.

tjkirch replied:
env_logger does have benefits, but I don't think it can be set up with split stdout/stderr like we need.  simplelog makes that easy.  That actually raises a point I missed here; metricdog service files should have `StandardError=journal+console` so we see errors on console, otherwise it'd be harder to debug.

webern replied:
Ugh, you're right, sadly `env_logger` won't work: https://docs.rs/env_logger/0.7.1/env_logger/struct.Builder.html#method.target and https://docs.rs/env_logger/0.7.1/env_logger/enum.Target.html

webern replied:
Switched to simplelog in [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737)

tjkirch commented on sources/metricdog/src/main.rs:
```diff
@@ -0,0 +1,121 @@
+    } else {
+        BottlerocketRelease::new()
+    }
+    .context(error::BottlerocketRelease)?;
+    let metricdog = Metricdog::from_parts(config, os_release, service_check)?;
+    match arguments.command {
```
```suggestion
    .context(error::BottlerocketRelease)?;
    
    let metricdog = Metricdog::from_parts(config, os_release, service_check)?;
    
    match arguments.command {
```
A little breathing room in this function?  :)

webern replied:
Aired out in [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737)

tjkirch commented on sources/metricdog/src/metricdog.rs:
```diff
@@ -0,0 +1,158 @@
+
+impl Metricdog {
+    /// Create a new instance by optionally passing in the `Config`, `BottlerocketRelease`, and
+    /// `ServiceCheck` objects. For each of these, if `None` is passed, then the default is used.
+    pub(crate) fn from_parts(
+        config: Config,
+        os_release: BottlerocketRelease,
+        healthcheck: Box<dyn ServiceCheck>,
```
I'm not sure I understand the comment, since none of these fields are `Option`..?

webern replied:
An example of "comments are lies." Fixed in [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

tjkirch commented on sources/metricdog/src/metricdog.rs:
```diff
@@ -0,0 +1,158 @@
+            q.append_pair("variant", &self.os_release.variant_id);
+            q.append_pair("arch", &self.os_release.arch);
+            q.append_pair("aws_region", &self.config.aws_region);
+            q.append_pair("seed", format!("{}", &self.config.seed).as_str());
```
```suggestion
            q.append_pair("seed", self.config.seed.to_string().as_str());
```
And maybe we don't need `as_str()`, given that `ignore_waves` below is passing a `String` directly?

webern replied:
Fixed in [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

tjkirch commented on sources/metricdog/src/service_check.rs:
```diff
@@ -0,0 +1,159 @@
+}
+
+impl Outcome {
+    fn is_exit_true(&self) -> bool {
```
Out of curiosity, why "true" here?

webern replied:
Two of the command-line calls use exit zero to represent true, so I guess I was trying to make the check for `== 0` to read more like a check for a truthy response.

tjkirch commented on sources/metricdog/src/service_check.rs:
```diff
@@ -0,0 +1,159 @@
+    let output = Command::new("systemctl")
+        .args(args)
+        .output()
+        .context(error::Command {
```
```suggestion
        .with_context(|| error::Command {
```
To avoid the iterator and copies in non-error cases.

webern replied:
Fixed in [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

tjkirch commented on sources/metricdog/src/service_check.rs:
```diff
@@ -0,0 +1,159 @@
+}
+
+const STATUS_PROPERTY: &str = "ExecMainStatus";
+const PROP_LEN: usize = STATUS_PROPERTY.len();
```
It feels like splitting on `=` might be simpler than length math?

webern replied:
Depends on whether you find functional or procedural code simpler, I suppose. But here is the split iterator version [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

tjkirch replied:
I don't think split has to be functional like it is in the current revision; I think you're implying that it's harder to read, and I would agree.  We can do something like [pluto](https://github.com/bottlerocket-os/bottlerocket/blob/741750ac742af91014c306183e94dc04285084f7/sources/api/pluto/src/main.rs#L208) or [bottlerocket-release](https://github.com/bottlerocket-os/bottlerocket/blob/741750ac742af91014c306183e94dc04285084f7/sources/bottlerocket-release/src/lib.rs#L75) that's a bit more procedural and clearer - `.next().context(bla)`

webern replied:
This seems readable [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737). Slightly different from the reference implementations since I'm checking that the key matches my expected key.

tjkirch reviewed (commented):
I think we should have the `-` prefix to not block startup, but otherwise these are basically nits.  Thanks!

webern commented:
> webern force-pushed the webern:healthdog branch from 153acff to b8b5c3a

Addresses some of the latest comments.

webern commented:
> webern force-pushed the webern:healthdog branch from b8b5c3a to c966f19

Addresses more PR comments. Everything is addressed except for the version bumping of the migration. Will do that after our next release and rebase then.

tjkirch commented on sources/api/schnauzer/src/lib.rs:
```diff
@@ -115,6 +115,7 @@ pub fn build_template_registry() -> Result<handlebars::Handlebars<'static>> {
     template_registry.register_helper("join_map", Box::new(helpers::join_map));
     template_registry.register_helper("default", Box::new(helpers::default));
     template_registry.register_helper("ecr-prefix", Box::new(helpers::ecr_prefix));
+    template_registry.register_helper("join-array", Box::new(helpers::join_array));
```
nit: `join_array` would be more consistent.  (I wish we had caught `ecr-prefix`.)

tjkirch commented on sources/metricdog/src/main.rs:
```diff
@@ -0,0 +1,127 @@
+
+fn main() -> ! {
+    let args = Arguments::from_args();
+    SimpleLogger::init(args.log_level, LogConfig::default()).expect("unable to configure logger");
```
What does `SimpleLogger` do for output?  Can we use `TermLogger` like we do elsewhere, so we can specify `TerminalMode::Mixed` to get errors on stderr?  I'm not sure the `StandardError` setting in `metricdog.service` will help us otherwise.

webern replied:
Bottom-line: `SimpleLogger` does what we want (splits stderr/stdout) and doesn't require an interactive terminal to be attached.

`SimpleLogger` [does split between stderr/stdout](https://github.com/Drakulix/simplelog.rs/blame/master/README.md#L13). I have tested and confirmed this.

`TermLogger` [requires an interactive terminal](https://github.com/Drakulix/simplelog.rs/blame/master/README.md#L14). @etungsten can speak to some difficulties with the update operator because `updog` uses `TermLogger`. [This issue](https://github.com/bottlerocket-os/bottlerocket-update-operator/issues/29#issue-580275364) is a result of using `TermLogger` which required the update operator to attach a TTY terminal.

tjkirch replied:
That's a good find.  We don't need/want the color support from `TermLogger` everywhere, so we could use `SimpleLogger` most places.  That may solve #576 (@jahkeup) and #697 (@etungsten) too.

tjkirch replied:
Created https://github.com/bottlerocket-os/bottlerocket/issues/1157 to track SimpleLogger use.

tjkirch commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -463,6 +475,74 @@ pub fn ecr_prefix(
+    for (i, value) in array.iter().enumerate() {
+        if i > 0 {
+            result.push_str(delimiter);
+            result.push(' ');
```
Hmm.  Should we force a space here, rather than the user specifying `", "` if they want a space?

tjkirch reviewed (approved):
Sorry, I wasn't trying to block this, I forgot to come back after the SimpleLogger discovery.  Just a couple suggestions above, and we need to update the version now that 1.0.2 is released.

webern commented:
> webern force-pushed the webern:healthdog branch from c966f19 to 7404d48

  * rename schnauzer helper from `join-array` to `join_array`
  * change schnauzer `join_array`helper to not add spaces
  * rebase
  * Change metrics URL to include the `v1/metrics` path, which the back-end expects and enforces.

Note: I wish I had done the rebase separately. I will remember to do rebases in their own push in the future to make diffing easier.

webern commented:
> webern force-pushed the webern:healthdog branch from c966f19 to 7404d48

Rebase, fix conflict in Cargo.toml

bcressey commented on packages/os/metricdog.timer:
```diff
@@ -0,0 +1,19 @@
+OnStartupSec=120
+# Run every 6 hours thereafter
+OnUnitActiveSec=21600
+# Don't fire at exactly the same second across machines started together.
+RandomizedDelaySec=10
```
We probably want more jitter than this, maybe as much as an hour.

webern replied:
Added jitter in this push [11b0927](https://github.com/bottlerocket-os/bottlerocket/compare/3db70e2f361b9cdd9306d75a1c2f6cf3d33eaadf..11b09274da45b6642d81f8164f2104b6157026e4)

webern replied:
Oops set jitter to 3600 this time for one hour [92230f5](https://github.com/bottlerocket-os/bottlerocket/compare/560f33678840b377569ea0dbb2d01633767172ae..92230f53e1f7aefca7c67130b785565a6d172824)

bcressey commented on sources/metricdog/README.md:
```diff
@@ -0,0 +1,55 @@
+* `event`: the event that invoked the report.
+* `version`: the Bottlerocket version.
+* `variant`: the Bottlerocket variant.
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
```
nit: `aarch64`

webern replied:
I missed this, need to fix.

webern replied:
Fixed in [560f336](https://github.com/bottlerocket-os/bottlerocket/compare/11b09274da45b6642d81f8164f2104b6157026e4..560f33678840b377569ea0dbb2d01633767172ae)

bcressey commented on sources/metricdog/src/metricdog.rs:
```diff
@@ -0,0 +1,158 @@
+use std::time::Duration;
+use url::Url;
+
+/// The send function optionally takes a timout parameter so that we can have a short timeout for
```
nit: typo, timeout

webern replied:
Fixed in [560f336](https://github.com/bottlerocket-os/bottlerocket/compare/11b09274da45b6642d81f8164f2104b6157026e4..560f33678840b377569ea0dbb2d01633767172ae)

bcressey commented on packages/os/metricdog.service:
```diff
@@ -0,0 +1,8 @@
+Type=oneshot
+RemainAfterExit=false
+StandardError=journal+console
+ExecStart=/usr/bin/metricdog send-health-ping
```
Can we set an overall timeout in the timer unit, or here (via `/usr/bin/timeout`)?

If the processes somehow get wedged in spite of our precautions, it would be good to prevent them from stacking up on the host.

webern replied:
I think we want TimeoutStartSec https://unix.stackexchange.com/a/599394/296877

webern replied:
[92230f5](https://github.com/bottlerocket-os/bottlerocket/compare/560f33678840b377569ea0dbb2d01633767172ae..92230f53e1f7aefca7c67130b785565a6d172824)

Added this here, will re-test just to make sure it still works.

webern replied:
OK I've done the testing for the timeout. It works! When I add a 40 second sleep, I see that the service fails with a timeout message. Remove that sleep and it works.

bcressey reviewed (approved):
Nicely done!

zmrow commented on sources/metricdog/README.md:
```diff
@@ -0,0 +1,55 @@
+* `version`: the Bottlerocket version.
+* `variant`: the Bottlerocket variant.
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
+* `aws_region`: the AWS region the machine is running in.
```
Does this need to be AWS specific?  As far as I can tell, we could just leave this as "region" and be a bit more agnostic so other platforms can make use of this field as well

tjkirch replied:
If it were running on another platform that had the same region name, we'd need some way to tell the difference, and I don't think the other fields would help.  Another option would be a "platform" field set to "aws", and a "region" field?

webern replied:
TL;DR: I agree `region` is better than `aws_region`, and I would prefer not to add a field such as `platform=aws`.

I initially called this field `region` and I strongly agree with changing it back to `region`. My thinking is that `region` is general-enough of a concept to be potentially useful with any cloud provider, and even for on-prem deployments, since it really just means 'general location'. In fact, I dislike calling `aws_region` so much that the current version of the backend accepts either `region` or `aws_region`! I should have spoken up sooner.

My disinclination to include `platform=aws` is less strong, but it feels unnecessary since the `variant` value conveys that information, and especially because we have not yet decided how we plan to codify such 'attributes of variant' in the system itself.

>  I don't think the other fields would help

Adding `platform=aws` is currently no different than `substr(variant, 3)`.

P.S.  Another way of stating my objection for `platform=aws` is, where does the value come from at runtime or compile time? Basically if variant in (aws-ecs-1, aws-k8s.1.15, aws-k8s.1.16, aws-k8s.1.17, aws-k8s.1.18) then `platform=aws`? Doesn't seem any cleaner to code that logic into the build/runtime of Bottlerocket right now than to do so when interpreting the metrics.

tjkirch replied:
I don't think variant will convey all of the information we need.  aws-dev can run anywhere, but may only be broken on one platform, for example.

zmrow replied:
Re: the region field - let's change it to be `region`; we all seem to agree on that.

Thinking about the `platform` field - I think it's a good idea, but it raises some questions like "How do we know what platform we're on?, etc"  

I don't think we need to block on this, and if we choose to ship without the changes, let's open an issue to track fixing this for other platforms.

zmrow reviewed (approved):
:lab_coat:

webern commented:
> webern force-pushed the webern:healthdog branch from a938b1f to 7c89fe2

Changes the name of the `aws_region` query parameter back to `region` (as discussed above).

webern commented:
>  webern force-pushed the webern:healthdog branch from 7c89fe2 to 518f4c6

Rebase.

webern commented:
> webern force-pushed the webern:healthdog branch from 518f4c6 to 520e55d

A substantial rebase.

bcressey commented:
metricdog should also respect `settings.network.https-proxy` and `settings.network.no-proxy`.

webern commented:
> metricdog should also respect `settings.network.https-proxy` and `settings.network.no-proxy`.

Maybe as simple as adding `.proxy(reqwest::Proxy::http("https://my.prox")?)` to the reqwest client when settings.network.https-proxy is set? I'm not sure I understand what settings.network.no-proxy means at the moment.

webern commented:
> webern force-pushed the webern:healthdog branch from 520e55d to b7a6dfe

rebase

webern commented:
> webern force-pushed the webern:healthdog branch from b7a6dfe to 10a32f7

rebase

webern commented:
> webern force-pushed the webern:healthdog branch from 10a32f7 to c6e68bf

Add migration to Release.toml

webern commented:
> webern force-pushed the webern:healthdog branch from c6e68bf to 3db70e2

Fixed the migration so it does the right things. I tested it through upgrade and downgrade, will add that to the PR notes.

webern commented:
> webern force-pushed the webern:healthdog branch from 3db70e2 to 11b0927

Increase jitter to one hour.

webern commented:
> webern force-pushed the webern:healthdog branch from 560f336 to 92230f5

- Correct the Jitter to 3600 (one hour)
- Add a TimeoutStartSec to the service

webern commented:
> webern force-pushed the webern:healthdog branch from 92230f5 to 312eff5

The last rebase?
